File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cpp
#include <string>
#include <bits/stdc++.h>
#include <vector>

using namespace std;

int solution(vector<string> friends, vector<string> gifts) {
    unordered_map<string, unordered_map<string,int>> m;
    unordered_map<string, int> m2;
    unordered_map<string, int> ans;

    for(int i = 0; i<friends.size(); i++){
        for(int j = 0; j<friends.size(); j++){
            if(i == j) continue;
            m[friends[i]][friends[j]] = 0;
        }
        m2[friends[i]] = 0; ans[friends[i]] = 0;
    }

    for(auto g: gifts){
        string given, received;
        stringstream s(g);
        s >> given >> received;

        m[given][received]++;
        m2[given]++;
        m2[received]--;
    }

//    for (const auto& pair1 : m) {
//        cout << pair1.first << " => { ";
//        for (const auto& pair2 : pair1.second) {
//            cout << pair2.first << ": " << pair2.second << " ";
//        }
//        cout << "}" << endl;
//    }


    int res = 0;

    for(auto i : friends){
        for(auto it : m[i]){
          if(m[i][it.first] > m[it.first][i]){
              ans[i]++;
              res = max(res, ans[i]);
          }else if(m[i][it.first] == m[it.first][i]){
              if(m2[i] > m2[it.first]) {
                  ans[i]++;
                  res = max(res, ans[i]);
              }
          }
        }
    }

    //cout << res;

    return res;
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js

/*
  A 와 B 가 선물을 주고 받았다면 더 많이 준 사람이 받는다
  두사람이 주고 받은 기록이 없다면 선물지수가 더 큰 사람이 받는다
  선물지수는 이번달 준 선물 개수 - 받은 선물

  선물 지수도 같다면 아무일도 이러나지 않는다
*/


function solution(friends, gifts) {
  var answer = 0;
  var past = {};
  var giftPoints = {};
  var toRecieve = {};
  

  for(let f of friends){ // 선물 주고받기 초기화
    if(!past[f]) past[f] = {};

    for(let f2 of friends){
      if(f === f2) continue;
      past[f][f2] = 0;
    }
    if(!giftPoints[f]) giftPoints[f] = 0;
    if(!toRecieve[f]) toRecieve[f] = 0;
  }

  for(let g of gifts){
    let [from, to] = g.split(' ');
    if(!past[from][to]) past[from][to] = 0;
    past[from][to] += 1;

    giftPoints[from]++;
    giftPoints[to]--;
  }


  for(let i = 0; i<friends.length - 1; i++){
    for(let j = i + 1; j<friends.length; j++){
      // 준 선물 수가 같다면
      if(past[friends[i]][friends[j]] === past[friends[j]][friends[i]]){
        if(giftPoints[friends[i]] !== giftPoints[friends[j]]){
          let larger = giftPoints[friends[i]] > giftPoints[friends[j]] ? friends[i] : friends[j];
          toRecieve[larger]++;
        }
      }
      else if(past[friends[i]][friends[j]] > past[friends[j]][friends[i]]){
        toRecieve[friends[i]]++
      }else{
        toRecieve[friends[j]]++;
      }
    }
  }


  for(let [key, val] of Object.entries(toRecieve)){
    answer = Math.max(val, answer);
  }
  return answer;
}